sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -738,6 +739,49 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelpe
 }
 
 /**
+ * Reorder associative integral-type operators and fold all constants into one.
+ */
+object ReorderAssociativeOperator extends Rule[LogicalPlan] {
+  private def isAssociativelyFoldable(e: Expression): Boolean =
+    e.deterministic && e.isInstanceOf[BinaryArithmetic] && e.dataType.isInstanceOf[IntegralType] &&
+      isSingleOperatorExpr(e)
+
+  private def isSingleOperatorExpr(e: Expression): Boolean = e.find {
+    case a: Add if a.getClass == e.getClass => false
+    case m: Multiply if m.getClass == e.getClass => false
+    case _: BinaryArithmetic => true
+    case _ => false
+  }.isEmpty
+
+  private def getOperandList(e: Expression): Seq[Expression] = e match {
+    case BinaryArithmetic(a, b) => getOperandList(a) ++ getOperandList(b)
+    case other => other :: Nil
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsDown {
```

Comment:
I see. That could be.
We also need to add `isSingleOperatorExpr` there.
Otherwise, `flattenAdd(Add(Multiply(1, 2), 3)) -> (3)`.